File: sources/java/VsamAccountOps.java
import com.rocketsoftware.jzos.*;
import java.util.Arrays;

/**
 * Demonstrates VSAM KSDS operations on the BankDemo customer dataset.
 *
 * Operations:
 *   LOOKUP  <custId>          - Locate and display a single customer by key
 *   BROWSE  <startKey> <count> - Browse records starting from a key
 *   UPDATE  <custId>          - Locate a customer and toggle the SendMail flag
 *
 * Dataset: MFI01V.MFIDEMO.BNKCUST (KSDS, key at offset 0, length 5)
 *
 * Usage via JVMLDM:
 *   PARM='... VsamAccountOps LOOKUP B0001'
 *   PARM='... VsamAccountOps BROWSE B0001 10'
 *   PARM='... VsamAccountOps UPDATE B0001'
 */
public class VsamAccountOps {

    // BNKCUST record layout (from CBANKVCS.CPY)
    private static final int REC_LEN = 250;
    private static final int PID_OFF = 0,        PID_LEN = 5;
    private static final int NAME_OFF = 5,       NAME_LEN = 25;
    private static final int NAMEFF_OFF = 30,    NAMEFF_LEN = 25;
    private static final int SIN_OFF = 55,       SIN_LEN = 9;
    private static final int ADDR1_OFF = 64,     ADDR1_LEN = 25;
    private static final int ADDR2_OFF = 89,     ADDR2_LEN = 25;
    private static final int STATE_OFF = 114,    STATE_LEN = 2;
    private static final int COUNTRY_OFF = 116,  COUNTRY_LEN = 6;
    private static final int POSTCODE_OFF = 122, POSTCODE_LEN = 6;
    private static final int TEL_OFF = 128,      TEL_LEN = 12;
    private static final int EMAIL_OFF = 140,    EMAIL_LEN = 30;
    private static final int SENDMAIL_OFF = 170, SENDMAIL_LEN = 1;
    private static final int SENDEMAIL_OFF = 171, SENDEMAIL_LEN = 1;

    private static final String SEPARATOR = "-".repeat(60);

    public static void main(String[] args) {
        try {
            run(args);
        } catch (Throwable t) {
            System.err.println("FATAL: " + t.getClass().getName() + ": " + t.getMessage());
            t.printStackTrace(System.err);
            System.exit(16);
        }
    }

    private static void run(String[] args) throws ZFileException {
        if (args.length < 1) {
            System.err.println("ERROR: Missing operation (LOOKUP, BROWSE, or UPDATE)");
            System.exit(12);
        }

        String op = args[0].toUpperCase();
        System.err.printf("Job: %s  Step: %s  Operation: %s%n",
            ZUtil.getCurrentJobname(), ZUtil.getCurrentStepname(), op);
        System.err.printf("DEBUG: args.length=%d  args=%s%n", args.length, Arrays.toString(args));

        switch (op) {
            case "LOOKUP":
                if (args.length < 2) {
                    System.err.println("ERROR: LOOKUP requires a customer ID argument");
                    System.exit(12);
                }
                doLookup(args[1]);
                break;
            case "BROWSE":
                String startKey = args.length > 1 ? args[1] : "";
                int count = args.length > 2 ? Integer.parseInt(args[2]) : 10;
                doBrowse(startKey, count);
                break;
            case "UPDATE":
                if (args.length < 2) {
                    System.err.println("ERROR: UPDATE requires a customer ID argument");
                    System.exit(12);
                }
                doUpdate(args[1]);
                break;
            default:
                System.err.println("ERROR: Unknown operation '" + op + "'");
                System.exit(12);
        }
    }

    // -------------------------------------------------------------------------
    // LOOKUP: Locate a single record by primary key (customer ID)
    // -------------------------------------------------------------------------

    private static void doLookup(String custId) throws ZFileException {
        System.out.println("=== VSAM LOOKUP ===");
        System.out.println("Searching for customer: '" + custId + "' (length=" + custId.length() + ")");
        System.out.println(SEPARATOR);

        System.err.printf("LOOKUP: custId='%s' len=%d bytes=%s%n",
            custId, custId.length(), Arrays.toString(custId.getBytes()));

        try (ZFile vsam = new ZFile("//DD:CUSTDATA", "type=record,rb")) {
            System.err.printf("LOOKUP: Opened. VsamType=%d  KeyLen=%d  LRECL=%d%n",
                vsam.getVsamType(), vsam.getVsamKeyLength(), vsam.getLrecl());

            byte[] key = makeKey(custId, vsam.getVsamKeyLength());
            System.err.printf("LOOKUP: key bytes=%s (len=%d)%n", Arrays.toString(key), key.length);

            boolean found = vsam.locate(key, ZFileConstants.LOCATE_KEY_EQ);
            System.err.printf("LOOKUP: locate(KEY_EQ) returned %b%n", found);

            if (found) {
                byte[] record = new byte[vsam.getLrecl()];
                int bytesRead = vsam.read(record);
                System.err.printf("LOOKUP: read() returned %d bytes%n", bytesRead);
                System.err.printf("LOOKUP: record[0..40]='%s'%n",
                    new String(record, 0, Math.min(40, bytesRead)));
                System.err.printf("LOOKUP: record hex[0..20]=%s%n", bytesToHex(record, 0, 20));
                if (bytesRead >= 0) {
                    printCustomerRecord(record);
                } else {
                    System.out.println("  Record not found (read returned " + bytesRead + ")");
                }
            } else {
                System.out.println("  Record not found for key: " + custId);
                // Try KEY_GE as fallback diagnostic
                boolean geFound = vsam.locate(key, ZFileConstants.LOCATE_KEY_GE);
                System.err.printf("LOOKUP: locate(KEY_GE) returned %b%n", geFound);
                if (geFound) {
                    byte[] record = new byte[vsam.getLrecl()];
                    int bytesRead = vsam.read(record);
                    System.err.printf("LOOKUP: GE read %d bytes, first 40='%s'%n",
                        bytesRead, new String(record, 0, Math.min(40, bytesRead)));
                }
            }
        }

        System.out.println(SEPARATOR);
    }

    // -------------------------------------------------------------------------
    // BROWSE: Sequential read starting from a key (KEY_GE)
    // -------------------------------------------------------------------------

    private static void doBrowse(String startKey, int maxCount) throws ZFileException {
        System.out.println("=== VSAM BROWSE ===");
        System.out.printf("Start key: '%s'  Max records: %d%n", startKey, maxCount);
        System.out.println(SEPARATOR);
        System.out.printf("  %-5s  %-25s  %-12s  %-30s  %s%n",
            "PID", "Name", "Phone", "Email", "Mail");
        System.out.println("  " + "-".repeat(80));

        try (ZFile vsam = new ZFile("//DD:CUSTDATA", "type=record,rb")) {
            System.err.printf("BROWSE: Opened. VsamType=%d  KeyLen=%d  LRECL=%d%n",
                vsam.getVsamType(), vsam.getVsamKeyLength(), vsam.getLrecl());
            if (!startKey.isEmpty()) {
                byte[] key = makeKey(startKey, vsam.getVsamKeyLength());
                System.err.printf("BROWSE: locate KEY_GE key=%s%n", Arrays.toString(key));
                boolean found = vsam.locate(key, ZFileConstants.LOCATE_KEY_GE);
                System.err.printf("BROWSE: locate returned %b%n", found);
            } else {
                byte[] key = new byte[vsam.getVsamKeyLength()];
                System.err.printf("BROWSE: locate KEY_FIRST key=%s%n", Arrays.toString(key));
                boolean found = vsam.locate(key, ZFileConstants.LOCATE_KEY_FIRST);
                System.err.printf("BROWSE: locate returned %b%n", found);
            }

            byte[] record = new byte[vsam.getLrecl()];
            int count = 0;

            while (vsam.read(record) >= 0 && count < maxCount) {
                if (count == 0) {
                    System.err.printf("BROWSE: first record hex[0..20]=%s%n", bytesToHex(record, 0, 20));
                    System.err.printf("BROWSE: first record text[0..40]='%s'%n",
                        new String(record, 0, Math.min(40, record.length)));
                }
                String pid = field(record, PID_OFF, PID_LEN);
                String name = field(record, NAME_OFF, NAME_LEN);
                String tel = field(record, TEL_OFF, TEL_LEN);
                String email = field(record, EMAIL_OFF, EMAIL_LEN);
                String sendMail = field(record, SENDMAIL_OFF, SENDMAIL_LEN);

                System.out.printf("  %-5s  %-25s  %-12s  %-30s  %s%n",
                    pid, name, tel, email, sendMail);
                count++;
            }

            System.out.println("  " + "-".repeat(80));
            System.out.printf("  Browsed %d records%n", count);
            System.err.printf("BROWSE: returned %d records from key '%s'%n", count, startKey);
        } catch (ZFileException e) {
            System.out.println("  No records found from key: " + startKey);
            System.err.println("BROWSE exception: " + e);
            e.printStackTrace(System.err);
        }

        System.out.println(SEPARATOR);
    }

    // -------------------------------------------------------------------------
    // UPDATE: Locate a record and toggle the SendMail flag
    // -------------------------------------------------------------------------

    private static void doUpdate(String custId) throws ZFileException {
        System.out.println("=== VSAM UPDATE ===");
        System.out.println("Updating customer: " + custId);
        System.out.println(SEPARATOR);

        System.err.printf("UPDATE: custId='%s' len=%d%n", custId, custId.length());

        try (ZFile vsam = new ZFile("//DD:CUSTDATA", "type=record,rb+")) {
            System.err.printf("UPDATE: Opened. VsamType=%d  KeyLen=%d  LRECL=%d%n",
                vsam.getVsamType(), vsam.getVsamKeyLength(), vsam.getLrecl());
            byte[] key = makeKey(custId, vsam.getVsamKeyLength());
            System.err.printf("UPDATE: key bytes=%s%n", Arrays.toString(key));
            boolean found = vsam.locate(key, ZFileConstants.LOCATE_KEY_EQ);
            System.err.printf("UPDATE: locate(KEY_EQ) returned %b%n", found);

            if (!found) {
                System.out.println("  Record not found for key: " + custId);
                return;
            }

            byte[] record = new byte[vsam.getLrecl()];
            int bytesRead = vsam.read(record);
            System.err.printf("UPDATE: read() returned %d bytes%n", bytesRead);
            System.err.printf("UPDATE: record[0..40]='%s'%n",
                new String(record, 0, Math.min(40, bytesRead >= 0 ? bytesRead : 0)));

            System.out.println("  Before update:");
            printCustomerRecord(record);

            // Toggle SendMail: "Y" <-> "N"
            String currentMail = field(record, SENDMAIL_OFF, SENDMAIL_LEN);
            String newMail = currentMail.equals("Y") ? "N" : "Y";
            System.arraycopy(newMail.getBytes(), 0, record, SENDMAIL_OFF, SENDMAIL_LEN);

            vsam.update(record, 0, record.length);
            System.err.printf("UPDATE: update() completed successfully%n");

            System.out.println("  After update:");
            printCustomerRecord(record);
            System.err.printf("UPDATE: Toggled SendMail: '%s' -> '%s'%n", currentMail, newMail);
        } catch (ZFileException e) {
            System.out.println("  Update failed: " + e.getMessage());
            System.err.println("UPDATE exception: " + e);
            e.printStackTrace(System.err);
        }

        System.out.println(SEPARATOR);
    }

    // -------------------------------------------------------------------------
    // Utilities
    // -------------------------------------------------------------------------

    private static void printCustomerRecord(byte[] record) {
        String pid = field(record, PID_OFF, PID_LEN);
        String name = field(record, NAME_OFF, NAME_LEN);
        String addr1 = field(record, ADDR1_OFF, ADDR1_LEN);
        String state = field(record, STATE_OFF, STATE_LEN);
        String postcode = field(record, POSTCODE_OFF, POSTCODE_LEN);
        String tel = field(record, TEL_OFF, TEL_LEN);
        String email = field(record, EMAIL_OFF, EMAIL_LEN);
        String sendMail = field(record, SENDMAIL_OFF, SENDMAIL_LEN);
        String sendEmail = field(record, SENDEMAIL_OFF, SENDEMAIL_LEN);

        System.out.printf("  PID: %s  Name: %s%n", pid, name);
        System.out.printf("  Address: %s, %s %s%n", addr1, state, postcode);
        System.out.printf("  Phone: %s  Email: %s%n", tel, email);
        System.out.printf("  SendMail: %s  SendEmail: %s%n", sendMail, sendEmail);
    }

    private static String field(byte[] record, int offset, int length) {
        return new String(record, offset, length).trim();
    }

    private static byte[] makeKey(String value, int keyLength) {
        byte[] key = new byte[keyLength];
        byte[] src = value.getBytes();
        System.arraycopy(src, 0, key, 0, Math.min(src.length, keyLength));
        return key;
    }

    private static String bytesToHex(byte[] data, int offset, int length) {
        StringBuilder sb = new StringBuilder();
        int end = Math.min(offset + length, data.length);
        for (int i = offset; i < end; i++) {
            sb.append(String.format("%02X ", data[i] & 0xFF));
        }
        return sb.toString().trim();
    }
}

File: sources/python/vsam_account_ops.py
"""
Step 4 Demo: Low-level VSAM Operations with the esos API.

Demonstrates direct VSAM KSDS operations on BankDemo datasets using the
low-level esos.esos API instead of the higher-level zoautil_py.
This is the Python equivalent of VsamAccountOps.java.

Four VSAM operations are shown:
  LOOKUP - Random keyed read: locate a specific customer by primary key
  BROWSE - Sequential browse: read records starting from a given key (KEY_GE)
  UPDATE - Read-for-update: locate a record, modify a field, rewrite it
  READ   - Sequential read: read all records from BNKACC (account dataset)

Key concepts demonstrated:
  - Low-level esos API: Esos.default.file_open(path, FileOptions)
  - FileOptions configuration: mode, recfm, lrecl, disposition, VSAM type
  - EsosFile as a context manager (with ... as f:)
  - VSAM locate options: KEY_EQ (exact match), KEY_GE (greater or equal)
  - Read-for-update: MODE_TYPE_READ | MODE_FLAG_UPDATE, then file.update()
  - Difference from zoautil_py: more control, but more setup required

When to use which API:
  - zoautil_py (zopen): Simple sequential read/write, less setup, familiar API
  - esos (EsosFile): Keyed access, update-in-place, browse, delete, position

Datasets:
  MFI01V.MFIDEMO.BNKCUST (VSAM KSDS, 250 bytes, key at offset 0) - LOOKUP/BROWSE/UPDATE
  MFI01V.MFIDEMO.BNKACC  (VSAM KSDS, 200 bytes, key at offset 5) - READ

See PYVSAM.jcl for the JCL that runs all four operations.

Usage:
  EXEC PGM=PYLDM,PARM='vsam_account_ops.py LOOKUP key'
  EXEC PGM=PYLDM,PARM='vsam_account_ops.py BROWSE start_key max_records'
  EXEC PGM=PYLDM,PARM='vsam_account_ops.py UPDATE key new_email'
  EXEC PGM=PYLDM,PARM='vsam_account_ops.py READ [max_records]'
"""

import sys

from esos.esos import (
    Esos,              # Main entry point: Esos.default gives the EsosContext
    EsosException,     # Exception raised on file I/O errors
    EsosFileMode,      # Read, Write, Append, Update flags
    EsosLocateOption,  # KEY_EQ, KEY_GE, KEY_FIRST, KEY_LAST, etc.
    EsosOpenFlags,     # OPEN_MODE_RECORD, OPEN_MODE_BINARY
    EsosDisposition,   # SHR, OLD, NEW, MOD (like JCL DISP=)
    EsosDsorg,         # Dataset organization: PS, VSAM, PO, etc.
    EsosVsamType,      # CLUSTER, PATH, AIX, etc.
    FileOptions,       # Structure holding all open parameters
)


# =============================================================================
# Record Layout: BNKCUST (from CBANKVCS.cpy)
# =============================================================================
# 250 bytes total, VSAM KSDS, primary key = bytes 0-4 (customer ID, 5 bytes)
#
# Offset  Length  Field             COBOL Definition
# ------  ------  -----             ----------------
#   0       5     Customer ID       PIC X(5)         [PRIMARY KEY]
#   5      25     Name              PIC X(25)
#  30      25     Name (formal)     PIC X(25)
#  55       9     SIN               PIC X(9)
#  64      25     Address line 1    PIC X(25)
#  89      25     Address line 2    PIC X(25)
# 114       2     State             PIC X(2)
# 116       6     Country           PIC X(6)
# 122       6     Post code         PIC X(6)
# 128      12     Phone             PIC X(12)
# 140      30     Email             PIC X(30)
# 170       1     Send mail flag    PIC X(1)
# 171       1     Send email flag   PIC X(1)
# 172       4     ATM PIN           PIC X(4)
# 176      74     Filler            PIC X(74)

CUST_LRECL = 250
CUST_PID = (0, 5)
CUST_NAME = (5, 25)
CUST_NAME_FF = (30, 25)
CUST_SIN = (55, 9)
CUST_ADDR1 = (64, 25)
CUST_ADDR2 = (89, 25)
CUST_STATE = (114, 2)
CUST_CNTRY = (116, 6)
CUST_POST = (122, 6)
CUST_TEL = (128, 12)
CUST_EMAIL = (140, 30)


def field(record, layout):
    """Extract a text field from a record buffer.

    COBOL PIC X fields are stored as fixed-width ASCII, padded with spaces.
    """
    offset, length = layout
    return bytes(record[offset:offset + length]).decode("ascii", errors="replace").strip()


def set_field(record, layout, value):
    """Set a text field in a mutable record buffer.

    Left-justifies and pads with spaces to fill the fixed field width,
    matching COBOL's MOVE semantics for PIC X fields.
    """
    offset, length = layout
    encoded = value.encode("ascii")[:length].ljust(length)
    record[offset:offset + length] = encoded


def open_custdata(update=False):
    """Open the CUSTDATA DD using the low-level esos API.

    This demonstrates configuring FileOptions manually — equivalent to
    the parameters you'd pass to ZFile in Java, but more explicit.

    Args:
        update: If True, opens for read+update (DISP=OLD).
                If False, opens for read-only (DISP=SHR).

    Returns:
        EsosFile instance (use as context manager: with open_custdata() as f:)
    """
    opts = FileOptions()

    # Mode: read-only or read+update
    # MODE_FLAG_UPDATE enables the file.update() method for rewriting records
    if update:
        opts.mode_flags = EsosFileMode.MODE_TYPE_READ | EsosFileMode.MODE_FLAG_UPDATE
    else:
        opts.mode_flags = EsosFileMode.MODE_TYPE_READ

    # Open flags: we want record-mode binary I/O (not text/stream)
    opts.open_flags = EsosOpenFlags.OPEN_MODE_RECORD | EsosOpenFlags.OPEN_MODE_BINARY

    # Record format: KS = Key-Sequenced (VSAM KSDS)
    opts.recfm = "KS"
    opts.lrecl = CUST_LRECL

    # Disposition: SHR for read-only, OLD for exclusive update access
    # This matches JCL DISP=SHR vs DISP=OLD
    opts.disposition = EsosDisposition.FLAG_DISP_SHR if not update else EsosDisposition.FLAG_DISP_OLD

    # Dataset organization and VSAM type
    opts.dsorg = EsosDsorg.VSAM
    opts.vsam_type = EsosVsamType.CLUSTER
    opts.vsam_key_length = 5  # Primary key is 5 bytes (customer ID)

    # Open the file. "//DD:CUSTDATA" refers to the DD allocated in JCL.
    # Esos.default is the singleton EsosContext (initialized by PYLDM).
    return Esos.default.file_open("//DD:CUSTDATA", opts)


def read_record(f):
    """Read one fixed-length record from the current VSAM position.

    Returns:
        bytearray of CUST_LRECL bytes, or None if at end-of-file.

    Note:
        Known issue in esos.py: EsosFile.read() raises EsosException on EOF
        instead of returning 0. The native esos_file_read() returns a
        non-zero status for VSAM status "10" (EOF), and raiseOnError()
        treats any non-zero return as an error. zoautil_py's readrecord()
        expects read() to return 0 at EOF (returning empty bytes), so
        this bug affects zoautil_py VSAM reads as well.
    """
    buf = bytearray(CUST_LRECL)
    try:
        n = f.read(buf, 0, CUST_LRECL)
        if n == 0:
            return None
        return buf
    except EsosException as e:
        # Workaround: esos.py raises EsosException on EOF instead of returning 0.
        # Check VSAM status codes to distinguish EOF from real errors:
        #   "10" = end of file (no more records)
        #   "23" = record not found (key doesn't exist)
        status = f.status()
        if status.code in ("10", "23"):
            return None
        raise


def print_customer(record):
    """Print a formatted customer record."""
    pid = field(record, CUST_PID)
    name = field(record, CUST_NAME)
    addr1 = field(record, CUST_ADDR1)
    state = field(record, CUST_STATE)
    post = field(record, CUST_POST)
    tel = field(record, CUST_TEL)
    email = field(record, CUST_EMAIL)
    print(f"  PID:    {pid}")
    print(f"  Name:   {name}")
    print(f"  Addr:   {addr1}")
    print(f"  State:  {state}  Post: {post}")
    print(f"  Tel:    {tel}")
    print(f"  Email:  {email}")


# =============================================================================
# LOOKUP: Random Read by Primary Key
# =============================================================================

def do_lookup(args):
    """Locate and display a single customer by exact key match.

    Uses KEY_EQ which finds only an exact match on the 5-byte primary key.
    If the customer doesn't exist, locate() returns False.
    """
    if len(args) < 1:
        print("Usage: vsam_account_ops.py LOOKUP <customer_id>", file=sys.stderr)
        return 1

    # Pad key to exactly 5 bytes (VSAM key must be full length)
    key = args[0].encode("ascii").ljust(5)[:5]
    print(f"=== VSAM LOOKUP: key='{key.decode()}' ===")

    # EsosFile is a context manager — automatically closes on exit
    with open_custdata() as f:
        # locate() positions the file to the matching record.
        # Returns False if no record matches (status "23").
        found = f.locate(key, EsosLocateOption.KEY_EQ)
        if not found:
            print(f"  Customer '{key.decode().strip()}' not found.")
            return 1

        # Read the record at the current position
        record = read_record(f)
        if record is None:
            print(f"  Customer '{key.decode().strip()}' not found (read failed).")
            return 1

        print_customer(record)

    print("=== Lookup Complete ===")
    return 0


# =============================================================================
# BROWSE: Sequential Read from a Starting Key
# =============================================================================

def do_browse(args):
    """Browse customers sequentially starting from a given key.

    Uses KEY_GE (greater than or equal) to find the starting position,
    then reads forward sequentially. This is the standard VSAM browse pattern.
    """
    # Default: start from beginning, show 10 records
    start_key = args[0].encode("ascii").ljust(5)[:5] if args else b'     '
    max_records = int(args[1]) if len(args) > 1 else 10

    print(f"=== VSAM BROWSE: start='{start_key.decode().strip()}', max={max_records} ===")
    print(f"{'PID':<6s} {'Name':<25s} {'State':<6s} {'Email':<30s}")
    print("-" * 70)

    with open_custdata() as f:
        # KEY_GE: position to first record with key >= start_key.
        # If start_key is all spaces, this effectively starts from the beginning.
        found = f.locate(start_key, EsosLocateOption.KEY_GE)
        if not found:
            print("  No records found at or after the given key.")
            return 0

        # Read forward sequentially from the positioned record
        count = 0
        while count < max_records:
            record = read_record(f)
            if record is None:
                break  # End of file
            pid = field(record, CUST_PID)
            name = field(record, CUST_NAME)
            state = field(record, CUST_STATE)
            email = field(record, CUST_EMAIL)
            print(f"  {pid:<5s} {name:<25s} {state:<6s} {email}")
            count += 1

    print("-" * 70)
    print(f"  {count} records displayed.")
    print("=== Browse Complete ===")
    return 0


# =============================================================================
# UPDATE: Read-for-Update and Rewrite
# =============================================================================

def do_update(args):
    """Locate a customer, modify the email field, and rewrite the record.

    This demonstrates the VSAM update pattern:
      1. Open with MODE_FLAG_UPDATE (and DISP=OLD for exclusive access)
      2. locate() the target record by key
      3. read() the record (this "locks" it for update)
      4. Modify the field(s) in the buffer
      5. update() rewrites the record at the same position

    The JCL must specify DISP=OLD on the DD for update access.
    """
    if len(args) < 1:
        print("Usage: vsam_account_ops.py UPDATE <customer_id> [new_email]",
              file=sys.stderr)
        return 1

    key = args[0].encode("ascii").ljust(5)[:5]
    new_email = args[1] if len(args) > 1 else ""
    print(f"=== VSAM UPDATE: key='{key.decode().strip()}', new_email='{new_email}' ===")

    # Open with update=True for read+update mode (DISP=OLD)
    with open_custdata(update=True) as f:
        found = f.locate(key, EsosLocateOption.KEY_EQ)
        if not found:
            print(f"  Customer '{key.decode().strip()}' not found.")
            return 1

        # Read the record — after this, file.update() will rewrite THIS record
        record = read_record(f)
        if record is None:
            print(f"  Customer '{key.decode().strip()}' not found (read failed).")
            return 1

        old_email = field(record, CUST_EMAIL)
        print(f"  Before: email='{old_email}'")

        # Modify the email field in our buffer
        set_field(record, CUST_EMAIL, new_email)

        # Rewrite the entire record at the same VSAM position
        f.update(bytes(record), 0, CUST_LRECL)

        print(f"  After:  email='{new_email}'")

    print("=== Update Complete ===")
    return 0


# =============================================================================
# READ Operation: Sequential VSAM Read of BNKACC
# =============================================================================

# BNKACC record layout (from CBANKVAC.cpy, VSAM KSDS, lrecl=200)
ACC_LRECL = 200
ACC_CUSTID = (0, 5)
ACC_ACCTID = (5, 9)
ACC_TYPE = (14, 1)


def open_accdata():
    """Open the ACCDATA DD (BNKACC dataset) for sequential VSAM reading.

    This demonstrates opening a second VSAM dataset with different
    parameters (LRECL=200, key length=9) compared to CUSTDATA (LRECL=250).
    """
    opts = FileOptions()
    opts.mode_flags = EsosFileMode.MODE_TYPE_READ
    opts.open_flags = EsosOpenFlags.OPEN_MODE_RECORD | EsosOpenFlags.OPEN_MODE_BINARY
    opts.recfm = "KS"
    opts.lrecl = ACC_LRECL
    opts.disposition = EsosDisposition.FLAG_DISP_SHR
    opts.dsorg = EsosDsorg.VSAM
    opts.vsam_type = EsosVsamType.CLUSTER
    opts.vsam_key_length = 9  # Account ID at offset 5, length 9

    return Esos.default.file_open("//DD:ACCDATA", opts)


def do_read(args):
    """Read BNKACC records sequentially using the esos VSAM API.

    Demonstrates VSAM sequential reading with the low-level esos API:
    explicit FileOptions, locate(KEY_FIRST), and read() in a loop.
    Compare with Step 2 (sequential_file_ops.py) which uses zoautil_py
    for non-VSAM sequential I/O.

    Args:
        args[0]: Number of records to display (default: 5)
    """
    display_n = int(args[0]) if args else 5
    print(f"=== VSAM READ: First {display_n} account records (esos API) ===")

    with open_accdata() as f:
        # Position to the very first record in key sequence
        f.locate(b'', EsosLocateOption.KEY_FIRST)

        buf = bytearray(ACC_LRECL)
        count = 0
        displayed = 0

        while True:
            try:
                n = f.read(buf, 0, ACC_LRECL)
                if n == 0:
                    break
            except EsosException:
                # Workaround: esos.py bug — EsosFile.read() raises
                # EsosException at EOF (VSAM status "10") instead of
                # returning 0. See read_record() docstring for details.
                break

            count += 1
            if displayed < display_n:
                cust_id = field(buf, ACC_CUSTID)
                account_id = field(buf, ACC_ACCTID)
                account_type = field(buf, ACC_TYPE)
                print(f"  Account: {account_id}  Customer: {cust_id}  Type: {account_type}")
                displayed += 1

        if displayed < count:
            print(f"  ... ({count - displayed} more records not shown)")
        print(f"  Total records: {count}")

    print("=== Read Complete ===")
    return 0


# =============================================================================
# Main Entry Point
# =============================================================================

def main(args=None):
    if args is None:
        args = sys.argv[1:]

    if len(args) < 1:
        print("Usage: vsam_account_ops.py LOOKUP|BROWSE|UPDATE|READ [args...]",
              file=sys.stderr)
        return 1

    mode = args[0].upper()
    remaining = args[1:]

    if mode == "LOOKUP":
        return do_lookup(remaining)
    elif mode == "BROWSE":
        return do_browse(remaining)
    elif mode == "UPDATE":
        return do_update(remaining)
    elif mode == "READ":
        return do_read(remaining)
    else:
        print(f"Unknown mode: {mode}. Use LOOKUP, BROWSE, UPDATE, or READ.",
              file=sys.stderr)
        return 1


if __name__ in ("__main__", "<run_path>"):
    main()
